README.md
<img alt="OrbTk" width="380" src="https://gitlab.redox-os.org/redox-os/assets/raw/master/logos/orbtk/logo_dark.png">

[![Build and test](https://github.com/redox-os/orbtk/workflows/CI/badge.svg)](https://github.com/redox-os/orbtk/actions)
[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](./LICENSE)
[![crates.io](https://img.shields.io/badge/crates.io-0.3.1alpha3-orange.svg)](https://crates.io/crates/orbtk/0.3.1-alpha3)
[![docs.rs](https://img.shields.io/badge/docs-0.3.1alpha3-blue.svg)](https://docs.rs/crate/orbtk/0.3.1-alpha3)

The Orbital Widget Toolkit is a cross-platform (G)UI toolkit for building scalable user interfaces with the programming language Rust. It's based
on the [Entity Component System Pattern](https://en.wikipedia.org/wiki/Entity_component_system) and provides a [functional Reactive](https://en.wikipedia.org/wiki/Functional_reactive_programming)-like API.

The main goals of OrbTk are speed, ease of use, and cross-platform compatibility.

<img alt="Calculator" width="350" src="https://gitlab.redox-os.org/redox-os/assets/raw/master/screenshots/Calculator.png">

## Features:

* Modern lightweight API
* Cross platform
* Modular crates
* Based on Entity Component System library [DCES](https://gitlab.redox-os.org/redox-os/dces-rust)
* Flexible event system
* Integrated widget library
* Custom widgets
* Custom theming engine
* Dynamic theme switching
* Integrated debugging tools

## Platforms

* Redox OS (native)
* Linux (native | cargo-node)
* macOS (native | cargo-node)
* Windows (native | cargo-node)
* openBSD (not tested, but should work)
* Web (cargo-node)
* Android (native planned | cargo-node)
* iOS (native planned | cargo-node planned)
* Ubuntu Touch (native planned | cargo-node planned)

## Planned features

* Conformable use of async
* More default widgets
* Book
* Animations
* Split application in modules
* 3D context
* More integrated debugging tools

## Usage

To include OrbTk in your project, add this dependency
line to your `Cargo.toml` file:

```text
orbtk = "0.3.1-alpha3"
```

To use the latest development version of OrbTk, add this dependency
line to your `Cargo.toml` file:

```text
orbtk = { git = "https://github.com/redox-os/orbtk.git", branch = "develop" }
```
You can also check out the OrbTk template project to start a new project: https://github.com/redox-os/orbtk_template.

## Minimal Example

```rust
use orbtk::prelude::*;

fn main() {
      Application::new()
        .window(|ctx| {
            Window::new()
                .title("OrbTk - minimal example")
                .position((100.0, 100.0))
                .size(420.0, 730.0)
                .child(TextBlock::new().text("OrbTk").build(ctx))
                .build(ctx)
        })
        .run();
}
```

## Base concepts

### Widget

Widgets are the building blocks of user interfaces in OrbTk. They are things like Buttons, TextBoxes, ListViews, Views (Screens) and Grid(Layout)s. Each widget implements the [Widget trait](https://github.com/redox-os/orbtk/blob/develop/crates/api/src/widget_base/mod.rs) and is generated by the [widget! macro](https://github.com/redox-os/orbtk/blob/develop/crates/api/src/macros.rs). A widget consists of a name like `Button` and a list of its properties like `text: String16`, `background: Brush` or `count: u32`. After the `build` method of a widget is called it's added to the Entity Component System where it exists as an `Entity` (index) with `Components`. The struct of a widget serves as a builder using the [builder pattern](https://doc.rust-lang.org/1.0.0/style/ownership/builders.html).

Basic usage of the widget! macro:

```rust
widget!(
    MyWidget {
      background: Brush,
      count: u32,
      text: String16,
      ...
    }
);
```

### Widget Templates

Each widget has to implement the [Template trait](https://github.com/redox-os/orbtk/blob/develop/crates/api/src/widget_base/template.rs). The template defines the default values of a widget's properties as well as its structure. A
Button e.g. consists of a `Container` widget, a `StackPanel` widget and a `TextBlock` widget.

Basic usage of the Template trait:

```rust
impl Template for MyWidget {
    fn template(self, id: Entity, ctx: &mut BuildContext) -> Self {
         self.name("MyWidget")
            .style("my_widget_style")
            .background("#000000")
            .count(0)
            .text("Initial text")
            .child(Container::new()
                    // Container references the same background as MyWidget
                    .background(id)
                    .child(TextBlock::new()
                            // TextBlock references the same text as MyWidget
                            .text(id)
                            .build(ctx)
                    )
                    .build(ctx)
            )
    }
}
```

### Widget State

The state of a widget is used to update its inner state. Each state has to implement the [State trait](https://github.com/redox-os/orbtk/blob/develop/crates/api/src/widget_base/state.rs). The inner state of a widget is represented by the 
current values of its properties.

Basic usage of the state trait:

```rust
#[derive(Default, AsAny)]
struct MyState {
    ...
}

impl State for MyState {
    fn update(&mut self, _: &mut Registry, ctx: &mut Context) {
        // update the widget
        ...
    }
}

widget!(
    // Add MyState as state of MyWidget
    MyWidget<MyState> {
        ...
    }
);
```

The [Context parameter](https://github.com/redox-os/orbtk/blob/develop/crates/api/src/widget_base/context.rs) of the update method provides access to the state's widget (entity) and its properties (components). It also provides functions to access the children of the widget, and to manipulate the widget tree.

### Styling widgets and define themes

OrbTk provides a theme engine base on [RON](https://github.com/ron-rs/ron). The engine provides the following features:

* Split theme in different files
* Outsource resources like colors and font stuff
* Derive styles
* Dynamic theme switch
* State styling (pressed | selected | focused | disabled)

Short example: 
```ron
Theme (
    styles: {
        "base": (
            properties: {
                "font_size": "$FONT_SIZE_12",
                "font_family": "$MEDIUM_FONT",
            }
        ),
        "button": (
            base: "base",
            properties: {
                "background": "$BLACK",
            },
            states: {
                "pressed": {
                    "background": "$WHITE",
                }
            }
        )
    },
    resource: {
        "BLACK": "#000000",
        "WHITE": "#ffffff",
        "MEDIUM_FONT": "Roboto-Medium",
        "FONT_SIZE_12": 12,
        "FONT_SIZE_16": 16,
    }
)
```

OrbTk will also provide a plain mechanism to style and theme widgets and UIs. 

## Run Examples

To build and run the examples you *will* need an C compiler (like `gcc`, `clang`, or MS's own compiler).

On Linux you also nee to install `cmake`. e.g.: 

```shell
sudo apt install cmake
```

If you have trouble to build and run the examples or you don't want to use a C compiler check the backend section please? It contains alternatives.

You can find examples in the `examples/` directory.

You can start the widgets example by executing the following command:

```text
cargo run --example widgets --release
```

OrbTk has integrated `debug` tools. If you want to show the bounds of all widgets (even invisible ones) and want to see a debug print of the whole widget tree, you can run the examples with `--features debug`, like this:

```text
cargo run --example widgets --release --features debug
```

## Run Examples with cargo-node

To run the examples as a browser, electron or cordova app you have to install cargo-node:

```text
cargo install -f cargo-node
```

Before you can use cargo-node you have to install `npm` version 6.9.0, which is included with `Node.js` version 10.16.3. You can download it from https://nodejs.org/dist/v10.16.3/. 

Rust's `cargo` is also required. The rest of cargo-node's dependencies are installed automatically.

### Start examples

You can run the "widgets" example by executing one of the following commands:

* Run as browser app:

```text
cargo node run --target browser --example widgets
```

* Run as electron app:

```text
cargo node run --target electron --example widgets
```

* Run as cordova app on android:

```text
cargo node run --target android --example widgets
```

## OrbTk backends

At the moment will evaluate different backends for OrbTk. A OrbTk backend consists of two different parts, window / events and rendering. OrbTk provides at the moment the following backends:

### orbraq

* default backend for Redox, Linux, macOS and Windows (default feature)
* window and events based on [OrbClient](https://gitlab.redox-os.org/redox-os/orbclient)
* 2D rendering based on [raqote](https://github.com/jrmuizel/raqote)
* Known issues: window does not redraw while resizing

### miniraq

* use with `miniraq` feature `cargo run --example widgets --features miniraq --no-default-features` 
* window and events based on [minifb](https://github.com/emoon/rust_minifb)
* 2D rendering based on [raqote](https://github.com/jrmuizel/raqote)
* Dependencies on Linux: libxkbcommon-dev, libwayland-cursor0, libwayland-dev
* Does not support all features of `orbraq` e.g. DropEvents, Clipboard access 

### glupath (experimental)

* use with `glupath` feature `cargo run --example widgets --features glupath --no-default-features`
* window and events based on [glutin](https://github.com/rust-windowing/glutin)
* 2D rendering based on [pathfinder](https://github.com/servo/pathfinder)
* Does not yet support all features of `orbraq` e.g. DropEvents, Clipboard access 

### stdweb

* default backend for web (default feature)
* window, events and 2D rendering based on [stdweb](https://github.com/koute/stdweb)
* Does not yet support all features of `orbraq` e.g. DropEvents, Clipboard access 

## Documentation

### Build and open documentation

You can build and view the latest documentation by executing the following command:

```text
cargo doc --no-deps --open
```

### OrbTk Manual

To build and run the latest version of the OrbTk manual check: [Manual](https://github.com/redox-os/orbtk/tree/develop/manual)

## Sub Crates

* [api](https://github.com/redox-os/orbtk/tree/develop/crates/api): base api elements of OrbTk e.g. widget and application parts
* [proc_macros](https://github.com/redox-os/orbtk/tree/develop/crates/proc_macros): procedural helper macros
* [render](https://github.com/redox-os/orbtk/tree/develop/crates/render): cross platform 2D/3D render library
* [shell](https://github.com/redox-os/orbtk/tree/develop/crates/api): cross platform window and event handling
* [theming](https://github.com/redox-os/orbtk/tree/develop/crates/theme): provide mechanism to style OrbTk UI's in rust and ron (replaces css-engine)
* [tree](https://github.com/redox-os/orbtk/tree/develop/crates/tree): tree structure based on DCES
* [utils](https://github.com/redox-os/orbtk/tree/develop/crates/utils): helper structs and traits
* [widgets](https://github.com/redox-os/orbtk/tree/develop/crates/widgets): base widget library

## Inspirations

* [Flutter](https://flutter.io/)
* [React](https://reactjs.org/)
* [Yew](https://github.com/DenisKolodin/yew)

## Showcases

* [Space Editor](https://codeberg.org/flovanco/space-editor): 2D Tile Map Editor compatible with OrbGame
* [PluralTaks](https://github.com/PluralTools/PluralTasks): Task app
* [OrbCalculator](https://gitlab.redox-os.org/redox-os/orbcalculator): Calculator based on OrbTk
* [Kanter](https://github.com/lukors/kanter): Node based texture editor 
* [twin-commander](https://github.com/kivimango/twin-commander): A twin-panel file manager specifically for the Redox OS 

## Contribution

If you want to help improve OrbTk you submit your feedback in the issue tracker, or make a pull request to fix an issue https://github.com/redox-os/orbtk/issues. You can also discuss OrbTk with us on the Redox chat https://redox-os.org/community/ (join the OrbTk channel).

#### Contribution check list

* Documentation for all `pub` structs, traits and funs
* Add tests if needed
* Use static &str for widget ids and new style definitions 
* For widget development check ProgressBar or Slider as example
* Add changes to changelog
* Expand examples or create a new one if needed
* `cargo fmt --all` add the end
* Create PR

## License

Licensed under MIT license ([LICENSE](LICENSE)).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.3.1 (OrbTk next) wip

### 0.3.1-alpha4 (wip)

* PasswordBox widget
* Clipboard service
* Text input copy and paste
* orbraq backend (Orbclient and raqote)
* Refactor on_changed callback, add key parameter
* Add support for linear gradients on themes
* Add radial gradients support
* Access RawWindowHandle from Context

### 0.3.1-alpha3

* Dynamic theme switch
* Add all material font icons as resource
* Replaces css-engine with custom Rust/Ron based theming
* Add widget access helpers for states
* API update check deprecated methods an replace to new ones
* Performance improvements
* Change state update order from tree order to incoming changes order
* NumericBox widget
* Update caret position on TextBox by mouse click
* Text input support for ', /, \, [, ], {, }
* Multiple window support (experimental)
* Pathfinder / Glutin backend (experimental)
* ProgressBar widget
* Measure distance between two Points
* Improve: Mouse event arguments
* Fix: Crash when a child widget is removed
* TabWidget widget
* Add on_changed property change callback to all widgets
* OrbTk book (manual) wip

### 0.3.1-alpha2

* ComboBox / ComboboxItem widget
* Slider widget
* Popup widget
* Overlay layer
* Service registry for states
* Settings service (serialize / deserialize data)
* Direct access of states in callbacks
* Impl RawWindowHandle for Context (wip)
* Sent requests to window shell
* Layout fixes and stack layout example
* Many web fixes
* State cleanup method
* Refactor setting of styling selectors
* TextBox select all (Ctrl + a)
* Text input support for !, @, #
* Borderless window

### 0.3.1-alpha1

* api crate: base api elements of OrbTk e.g. widget and application parts
* css-engine crate: parse and read values from a css file
* proc_macros crate: procedural helper macros
* render crate: cross platform 2D/3D render library
* shell crate: cross platform window and event handling
* theme crate: OrbTks default theme (light and dark)
* tree crate: tree structure based on DCES
* utils crate: helper structs and traits
* widgets crate: base widget library
* Button widget
* Canvas widget
* CheckBox widget
* Container widget
* Cursor widget
* FontIconBlock widget
* Image widget
* Items widget
* ListView widget
* ScrollBar widget
* ScrollIndicator widget
* ScrollViewer widget
* Stack widget
* Switch widget
* TextBlock widget
* TextBox widget
* ToggleButton widget
* Window widget


examples/canvas.rs
use orbtk::prelude::*;
use std::cell::Cell;

use euc::{buffer::Buffer2d, rasterizer, Pipeline};
use vek::*;

// Cube is copied from euc spinning_cube example
// https://github.com/zesterer/euc/blob/master/examples/spinning_cube.rs
struct Cube<'a> {
    mvp: Mat4<f32>,
    positions: &'a [Vec4<f32>],
}

impl<'a> Pipeline for Cube<'a> {
    type Vertex = (usize, Rgba<f32>);
    type VsOut = Rgba<f32>;
    type Pixel = u32;

    #[inline(always)]
    fn vert(&self, (v_index, v_color): &Self::Vertex) -> ([f32; 4], Self::VsOut) {
        ((self.mvp * self.positions[*v_index]).into_array(), *v_color)
    }

    #[inline(always)]
    fn frag(&self, v_color: &Self::VsOut) -> Self::Pixel {
        let bytes = v_color.map(|e| (e * 255.0) as u8).into_array();
        (bytes[2] as u32)
            | (bytes[1] as u32) << 8
            | (bytes[0] as u32) << 16
            | (bytes[3] as u32) << 24
    }
}

#[derive(Clone, Default, PartialEq, Pipeline)]
struct CubePipeline {
    spin: Cell<f32>,
}

impl RenderPipeline for CubePipeline {
    fn draw(&self, render_target: &mut RenderTarget) {
        if render_target.width() == 0. || render_target.height() == 0. {
            return;
        }

        let mut color = Buffer2d::new(
            [
                render_target.width() as usize,
                render_target.height() as usize,
            ],
            0,
        );
        let mut depth = Buffer2d::new(
            [
                render_target.width() as usize,
                render_target.height() as usize,
            ],
            1.0,
        );

        let mvp = Mat4::perspective_fov_rh_no(
            1.3,
            render_target.width() as f32,
            render_target.height() as f32,
            0.01,
            100.0,
        ) * Mat4::translation_3d(Vec3::new(0.0, 0.0, -2.0))
            * Mat4::<f32>::scaling_3d(0.4)
            * Mat4::rotation_x((self.spin.get() * 0.002) * 8.0)
            * Mat4::rotation_y((self.spin.get() as f32 * 0.004).cos() * 4.0)
            * Mat4::rotation_z((self.spin.get() as f32 * 0.008).sin() * 2.0);

        Cube {
            mvp,
            positions: &[
                Vec4::new(-1.0, -1.0, -1.0, 1.0), // 0
                Vec4::new(-1.0, -1.0, 1.0, 1.0),  // 1
                Vec4::new(-1.0, 1.0, -1.0, 1.0),  // 2
                Vec4::new(-1.0, 1.0, 1.0, 1.0),   // 3
                Vec4::new(1.0, -1.0, -1.0, 1.0),  // 4
                Vec4::new(1.0, -1.0, 1.0, 1.0),   // 5
                Vec4::new(1.0, 1.0, -1.0, 1.0),   // 6
                Vec4::new(1.0, 1.0, 1.0, 1.0),    // 7
            ],
        }
        .draw::<rasterizer::Triangles<_, rasterizer::BackfaceCullingEnabled>, _>(
            &[
                // -x
                (0, Rgba::green()),
                (3, Rgba::blue()),
                (2, Rgba::red()),
                (0, Rgba::green()),
                (1, Rgba::red()),
                (3, Rgba::blue()),
                // +x
                (7, Rgba::blue()),
                (4, Rgba::green()),
                (6, Rgba::red()),
                (5, Rgba::red()),
                (4, Rgba::green()),
                (7, Rgba::blue()),
                // -y
                (5, Rgba::blue()),
                (0, Rgba::red()),
                (4, Rgba::green()),
                (1, Rgba::green()),
                (0, Rgba::red()),
                (5, Rgba::blue()),
                // +y
                (2, Rgba::red()),
                (7, Rgba::blue()),
                (6, Rgba::green()),
                (2, Rgba::red()),
                (3, Rgba::green()),
                (7, Rgba::blue()),
                // -z
                (0, Rgba::red()),
                (6, Rgba::green()),
                (4, Rgba::blue()),
                (0, Rgba::red()),
                (2, Rgba::blue()),
                (6, Rgba::green()),
                // +z
                (7, Rgba::green()),
                (1, Rgba::red()),
                (5, Rgba::blue()),
                (3, Rgba::blue()),
                (1, Rgba::red()),
                (7, Rgba::green()),
            ],
            &mut color,
            Some(&mut depth),
        );

        render_target.draw(color.as_ref());
    }
}

// OrbTk 2D drawing
#[derive(Clone, Default, PartialEq, Pipeline)]
struct Graphic2DPipeline;

impl RenderPipeline for Graphic2DPipeline {
    fn draw(&self, render_target: &mut RenderTarget) {
        let mut render_context =
            RenderContext2D::new(render_target.width(), render_target.height());

        let width = 120.0;
        let height = 120.0;

        let x = (render_target.width() - width) / 2.0;
        let y = (render_target.height() - height) / 2.0;
        // render_context.set_fill_style(utils::Brush::SolidColor(Color::from("#000000")));

        render_context.set_fill_style(utils::Brush::Gradient(Gradient {
            kind: GradientKind::Linear(LinearGradientCoords::Ends {
                start: Point::new(x, y),
                end: Point::new(x + width, y + height),
            }),
            stops: vec![
                GradientStop {
                    pos: 0.0,
                    color: Color::from("#0021EB"),
                },
                GradientStop {
                    pos: 0.5,
                    color: Color::from("#CE2F24"),
                },
                GradientStop {
                    pos: 1.0,
                    color: Color::from("#70EF49"),
                },
            ],
            repeat: false,
        }));
        render_context.fill_rect(x, y, width, height);
        // render_target.draw(render_context.data());
    }
}

#[derive(Default, AsAny)]
pub struct MainViewState {
    cube_spin: f32,
}

impl MainViewState {
    fn spin(&mut self) {
        self.cube_spin += 32.0;
    }
}

impl State for MainViewState {
    fn update(&mut self, _: &mut Registry, ctx: &mut Context) {
        if let Some(cube) = ctx
            .widget()
            .get_mut::<DefaultRenderPipeline>("render_pipeline")
            .0
            .as_any()
            .downcast_ref::<CubePipeline>()
        {
            cube.spin.set(self.cube_spin);
        }
    }
}

widget!(
    MainView<MainViewState> {
         render_pipeline: DefaultRenderPipeline
    }
);

impl Template for MainView {
    fn template(self, id: Entity, ctx: &mut BuildContext) -> Self {
        self.name("MainView")
            .render_pipeline(DefaultRenderPipeline(Box::new(CubePipeline::default())))
            .child(
                Grid::new()
                    .rows(Rows::create().push("auto").push("*").push("auto").push("*"))
                    .child(
                        TextBlock::new()
                            .attach(Grid::row(0))
                            .text("Canvas (render with euc crate)")
                            .style("text-block")
                            .style("text_block_header")
                            .margin(4.0)
                            .build(ctx),
                    )
                    .child(
                        Canvas::new()
                            .attach(Grid::row(1))
                            .render_pipeline(id)
                            .build(ctx),
                    )
                    .child(
                        Button::new()
                            .text("spin cube")
                            .v_align("end")
                            .attach(Grid::row(1))
                            .margin(4.0)
                            .on_click(move |states, _| {
                                states.get_mut::<MainViewState>(id).spin();
                                true
                            })
                            .build(ctx),
                    )
                    .child(
                        TextBlock::new()
                            .attach(Grid::row(2))
                            .text("Canvas (render with OrbTk)")
                            .style("text-block")
                            .style("text_block_header")
                            .margin(4.0)
                            .build(ctx),
                    )
                    .child(
                        Canvas::new()
                            .attach(Grid::row(3))
                            .render_pipeline(DefaultRenderPipeline(Box::new(
                                Graphic2DPipeline::default(),
                            )))
                            .build(ctx),
                    )
                    .build(ctx),
            )
    }
}

fn main() {
    // use this only if you want to run it as web application.
    orbtk::initialize();

    Application::new()
        .window(|ctx| {
            Window::new()
                .title("OrbTk - canvas example")
                .position((100.0, 100.0))
                .size(420.0, 730.0)
                .resizeable(true)
                .child(MainView::new().build(ctx))
                .build(ctx)
        })
        .run();
}
